codex/merkletree/merkletree.nim

```diff
@@ -0,0 +1,111 @@
+import std/sequtils
+import std/math
+import std/bitops
+import std/sugar
+
+import pkg/libp2p
+import pkg/stew/byteutils
+
+import pkg/questionable
+import pkg/questionable/results
+
+
+type
+  MerkleHash* = MultiHash
+  MerkleTree* = ref object of RootObj
+    nodes: seq[seq[MerkleHash]]
+  MerkleTreeBuilder* = ref object of RootObj
+    mcodec: ?MultiCodec
+    leaves: seq[MerkleHash]
+  MerkleProof* = ref object of RootObj
+    path: seq[MerkleHash]
+
+proc `$`*(t: MerkleTree): string =
+  result &= "height: " & $t.nodes.len
+  result &= "\nnodes: " & $t.nodes
+
+proc addLeaf*(self: MerkleTreeBuilder, hash: MerkleHash): ?!void =
```

Comment:
This could just be called `add`. It's pretty common in Nim to overload common operations that have similar semantics like `len`.

Reply:
I removed the builder altogether and replaced it with a single proc `newTree`